### DISCUSSION TITLE
Translations. What is the strategy?

### DISCUSSION THREAD

**Discussion by joeyrogues** [strapi/strapi#7284] [Legacy Topics]
Hello everyone :smiley_cat: 

Out of curiosity I ran the following code in several /translations directories

```js
const en = require("./en.json");
const fr = require("./fr.json");
const uk = require("./uk.json");
// ...

const allKeys = [
  ...new Set(
    [en, fr, uk]
      .map(Object.keys)
      .flat()
      .sort()
  )
];

Object.entries({ fr, uk })
  .map(([lang, trads]) => [lang, Object.keys(trads)])
  .forEach(([lang, langKeys]) => {
    const translated = allKeys.filter((k) => langKeys.includes(k))
    console.log(`${lang}: ${translated.length}/${allKeys.length} (${allKeys.length - translated.length} missing)`)
  });
```

I saw that many translations are missing. Which is not a big deal per se.

My question is: What is the overall strategy in terms of translations.
Can I (for instance):
1. create a test that checks all the keys are available for every language?
2. add null values for missing translations so we can visually see which translations are missing?
3. enforce alphabetical order for the translations keys (maybe)?

Let me know what you think.

BTW/Unrelated: so far I love the solution. It has a lot of value.

Thank you :smile_cat:

**Comment by lauriejim** [comment]
Hello! thank you for this question.
I will let @soupette give you the answer to that.

**Comment by soupette** [comment]
Hello!

Here are some answers regarding your questions:

> 1. create a test that checks all the keys are available for every language?

This is possible it would be a great addition. There's already a script that removes all the old keys in every translation file that is not `en.json`. Here's the strategy: the front-end core team maintains the translations of every `en.json` files and the script removes the uneeded keys in the others one. It could be easy to extend this script to know which keys are missing in the other languages.

> add null values for missing translations so we can visually see which translations are missing?

This won't be a good idea as there's a fallback to en if some keys are missing in a language. If we were to add `null` or empty script values there won't be any text displayed in the other language which is a behaviour we don't want in the administration panel. 

> enforce alphabetical order for the translations keys (maybe)?

The script mentioned in the first answer does this. If think we could just add it to the CI so it is handled in each translation PR.
